_source/_posts/2020-11-11-the-development-environment-of-the-future.md

```diff
@@ -0,0 +1,279 @@
+description: "A glimpse into the near future of programming"
+tags: [software]
+tweets:
+- "What will software development look like in the future? Okta Developer Advocate @jf has some ideas."
+- "Explore how programmers will work in the future with Okta Developer Advocate @jf"
+image: blog/the-development-environment-of-the-future/card.png
+type: awareness
+---
+
+Here's a thought exercise for you: How will we develop software in the near
+future? Below, I'll lay out what I think that looks like. Some of the things in
+this post exist in some form already.  And some things don't exist... yet. 
+
+To begin, I'm going to start with a short story about someone using this
+futuristic development environment:
+
+## A Story
+
+Imagine Viola. She is a software developer at Sunnyvale Systems, a logistics
+company. Viola is on the phone with a customer, who called in to report a bug in
+their route planning software. The customer is Polish, and unable to type the
+Polish letter "Ś" into text fields in the Sunnyvale Systems user interface. It
+takes a little while to figure out what the bug is, but Viola is able to get to
+the bottom of the issue by connecting her development environment directly to
+the customer's  browser session.
+
+Viola tries a few things with the customer on the line. "Try now," she says, and
+the customer tries again. No dice. A few more changes, and Viola asks the
+customer to try  again, repeating this until she finds the solution—the
+Sunnyvale Systems UI ignores the "Ctrl+S" keyboard combination to prevent the
+"Save webpage" dialog  from opening for people who are used to Microsoft Word. 
+
+However, Polish keyboards send "Ctrl+Alt+S" when a user types "Right Alt+S" to
+enter the "Ś" letter, and the logic to ignore "Ctrl+S" interferes with this.
+[The fix is to only ignore "Ctrl+S" if "Alt" isn't pressed as well](https://medium.engineering/the-curious-case-of-disappearing-polish-s-fa398313d4df).
+
+Once Viola figures out what the issue is, it's time to move her changes to the
+production system. To do this, Viola points her development environment at the
+production system, makes the change, but gets a warning that her updated change
+has no tests for it. So she quickly adds a test to make sure that the Ctrl+Alt+S
+key combination is not ignored. About 500 milliseconds after that test passes,
+her change is live. She asks the customer to reload his app one more time and
+try typing "Ś" —it works, and the customer hangs up, 
+[a little shocked that his issue was fixed so quickly](http://www.paulgraham.com/road.html).
+
+## Why Isn't the "Future" Here Already?
+
+Computers have barely achieved even a fraction of their potential. Most of what
+is holding back the potential of computers is not the technology itself, it's
+the mental models that humans use to interact with computers.
+
+Even with all of the advances that we've made since
+[1981](https://en.wikipedia.org/wiki/Xerox_Star#The_Xerox_Alto), the way that we
```

Comment:
The only thing I'd suggest is to NEVER use Wikipedia as a link. Because it's publicly editable, it may not say the same thing tomorrow as it does today. Much less in a year.